_drafts/ota-delta-updates.md

```diff
@@ -0,0 +1,600 @@
+delta updates allow us to do.
+
+The process is broken up into two steps. First, we need to extract the bits that
+have changed from our new image, we'll call it DIFF.
+
+<img width=600px src="{% img_url fwup-delta/new-code-DIFF.png %}" />
+
+Then we reconstruct our image from our original firmware and our delta image,
+which we'll call PATCH.
+
+<img width=700px src="{% img_url fwup-delta/new-code-PATCH.png %}" />
+
+### Pros and Cons of Delta Updates
+
+The obvious advantage of delta updates is the small size of the resulting image.
+Delta images are often one to two orders of magnitude smaller than full system
+images. The reduction in size has multiple beneficial effects:
+
+1. OTA updates become possible over very low bandwidth links (e.g. LoRaWAN).
+2. Updates are much faster over low bandwidth links. For example, a 10MB image
+   may take over 15 minutes to download over a BLE connection to a mobile phone,
+   even at [peak throughput]({% post_url 2019-09-24-ble-throughput-primer %}). A
+   delta update would take less than 1 minute to download, leading to a much
+   better customer experience and less risk of power loss mid-update.
+3. The lifetime of flash memory can be extended, as fewer writes are needed to
+   install a delta image than a full image.
+4. OTA consumer less power, thanks to the reduced communication and flash
+   writing required.
+
+However, there is no such thing as a free lunch! Delta updates come with some
+downsides worth considering:
+
+1. More complexity! Your firmware update code now needs to implement the
+   patching algorithm, which creates more opportunities for bugs.
+2. Delta updates requires more code space to implement the patching algorithm,
+   so if you are constrained already it may not be appropriate for your project.
+3. Managing delta updates creates more complexity on the backend and on the
+   release process.
+
+That last point bears elaborating a bit. Unlike all system images, each delta
+image is only compatible with a single original version. In other words, while a
+system image for firmware v1.2.0 can upgrade any prior version to v1.2.0, a
+delta update for "v1.1.0→v1.2.0" can only upgrade devices running v1.1.0 to
+v1.2.0.
+
+This means that your OTA backend needs to be sophisticated enough to present
+delta updates when devices are running compatible versions, and in all other
+cases present a full system update. And each firmware releases requires you to
+compile and upload a number of delta images for your versions in the field.
+
+> Our OTA backend at [Memfault](https://memfault.com) supports delta updates!
+> You can read more about in our
+> [documentation](https://docs.memfault.com/docs/platform/ota/#delta-releases).
+
+### Jojodiff
+
+One of the key components of a delta update system is a binary diff and patch
+system. There are remarkably few libraries that provide this functionality. The
+excellent BSDiff[^bsdiff], and XDelta[^xdelta] both require too much memory to
```

Comment:
These algorithms don't require loading the patched binary into memory like they say on the box, though, and can work just fine without much memory at all (source: we used both at Fitbit). You do I think have to tweak the implementations a tiny bit to handle that though.

Reply:
Added a note.